### PR TITLE
GLCORE: some performance optimizations

### PR DESCRIPTION
**Pull request by 0-8-15**
Remove use of `length`, avoid repeated list traversals and
combinations of apply+append.

### CODE REVIEW

File: modules/ln_glcore/glcore.scm [hunk 1]
```diff
@@ -226,19 +241,24 @@ SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
...
-    (set! glcore:clipx1 (car (car newlist)))
-    (set! glcore:clipy1 (cadr (car newlist)))
-    (set! glcore:clipx2 (caddr (car newlist)))
-    (set! glcore:clipy2 (cadddr (car newlist)))
-  ))  
-  (set! glcore:cliplist newlist)
- ))
+         (newcoords
+          (if (fx= (length coords) 4)
+              (map lambdanative#flo
```
Comment (mgorges) [inline modules/ln_glcore/glcore.scm:246]: `lambdanative#flo` is not defined, which causes a segfault in `glgui-list` - but what concerns me more is that I don't get a warning during compilation for it?

**Comment by 0-8-15** [comment]
Am Mon, 02 Nov 2020 01:55:58 -0800
schrieb Matthias Görges <notifications@github.com>:

> @mgorges commented on this pull request.
>
> `lambdanative#flo` is not defined, which causes a segfault in
> `glgui-list` - but what concerns me more is that I don't get a
> warning during compilation for it?

Sorry.  I overlooked this one, because I forgot that I rejected the
part of the related patch, which ... I can't find anymore, buried in
the update/pull history.

The `lambdanative#flo` binding was originally introduced as a
*procedural* binding to the current implementation of `flo` as a
measurement to enable backward compatibility for future, source code
compatible redefinition of `flo` as a macro. --- Since it is not
possible in Gambit to have both at the same time.  Or at least I don't
know hot to do that.

Another change in my queue here, which does exactly that is already
active in my tree.

I do not have the resources to test a full cycle to undo and then redo
this change.

Can you still figure this delta and revert it?

File: modules/ln_glcore/glcore.scm [hunk 2]
```diff
@@ -180,18 +190,23 @@ SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
...
-          ((fx= (u8vector-length data) (* 4 w h)) GL_RGBA)
-          (else (log-error "glCoreTextureCreate: Invalid data range") #f)))
-        (interpolation (if (>= (length aux) 1) (car aux) GL_LINEAR))
-        (wrap (if (>= (length aux) 2) (cadr aux) GL_CLAMP)))
-    (table-set! glCore:textures idx (##still-copy 
-       (vector #f (u32vector 0) w h (##still-copy data) pixeltype interpolation wrap)))
-    (set! glCore:tidx (fx+ glCore:tidx 1))
-    idx))
+  (let* ((o1x (pair? aux))
+         (o2 (and o1x (cdr aux))))
```
Comment (mgorges) [inline modules/ln_glcore/glcore.scm:194]: Note to self: the `(and o1x (cdr aux))` returns `#f` or `(cdr aux)`

File: modules/ln_glcore/glcore.scm [hunk 3]
```diff
@@ -226,19 +241,24 @@ SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
...
-            (if (null? oldlist) oldlist (cdr oldlist))))) 
-  (if (not (null? newlist)) (begin
-    (set! glcore:clipx1 (car (car newlist)))
-    (set! glcore:clipy1 (cadr (car newlist)))
-    (set! glcore:clipx2 (caddr (car newlist)))
-    (set! glcore:clipy2 (cadddr (car newlist)))
-  ))  
-  (set! glcore:cliplist newlist)
- ))
+         (newcoords
```
Comment (mgorges) [inline modules/ln_glcore/glcore.scm:244]: Note to self - none of these are changes, its purely reformatting code.

File: modules/ln_glcore/glcore.scm [hunk 4]
```diff
@@ -250,15 +270,22 @@ SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
...
-         (apply glCore:TextureDrawUnClipped (append (list (flo x) (flo y) w h t (flo x1) (flo y1) (flo x2) (flo y2) (flo r)) 
-           (if (null? colors) '() (car colors))))
-         (apply glCore:TextureDrawClipped (append (list (flo x) (flo y) w h t (flo x1) (flo y1) (flo x2) (flo y2) (flo r)) 
-           (if (null? colors) '() (car colors)))))
-      ) (log-error "glCoreTextureDraw: unbound index " t))
-  ))
+        (let ((w (flo (if (fx= (fix w0) 0) (vector-ref entry 2) w0)))
+              (h (flo (if (fx= (fix h0) 0) (vector-ref entry 3) h0))))
+          (if (null? glcore:cliplist)
+              (if (pair? colors)
```
Comment (mgorges) [inline modules/ln_glcore/glcore.scm:276]: Note to self - consider reformatting to 2 spaces as everywhere else before merging

File: modules/ln_glcore/glcore.scm [hunk 5]
```diff
@@ -309,22 +336,24 @@ SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
...
-          (glCoreColor (car colors))
-          (glCoreVertex2f (fl- cw2) ch2 c@x1 c@y2)
-          (glCoreColor (cadr colors))
-          (glCoreVertex2f cw2 ch2 c@x2 c@y2)
-          (glCoreColor (caddr colors))
-          (glCoreVertex2f (fl- cw2) (fl- ch2) c@x1 c@y1)
-          (glCoreColor (cadddr colors))
-          (glCoreVertex2f cw2 (fl- ch2) c@x2 c@y1)
-        ))
+        (if (null? colors)
```
Comment (mgorges) [inline modules/ln_glcore/glcore.scm:339]: Note: None of these are changes - all is reformatting

File: modules/ln_glcore/glcore.scm [hunk 6]
```diff
@@ -309,22 +336,24 @@ SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
...
+            (begin
+              (glCoreVertex2f (fl- cw2) ch2 c@x1 c@y2)
+              (glCoreVertex2f cw2 ch2 c@x2 c@y2)
+              (glCoreVertex2f (fl- cw2) (fl- ch2) c@x1 c@y1)
+              (glCoreVertex2f cw2 (fl- ch2) c@x2 c@y1)
+              )
+            (let ((colors (list->vector colors)))
+              ;; TODO: color interpolation here!
```
Comment (mgorges) [inline modules/ln_glcore/glcore.scm:347]: Changes here vector->list

File: modules/ln_glcore/glcore.scm [hunk 7]
```diff
@@ -338,24 +367,26 @@ SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
...
-        (glRotatef r 0. 0. 1.)
-        (_glCoreTextureBind t)
-        (glCoreBegin GL_TRIANGLE_STRIP)
-        (for-each (lambda (p) 
-            (let* ((x (fl- (car p) cx))
-                   (y (fl- (cadr p) cy))
-                   (tx (caddr p))
-                   (ty (cadddr p)))
-              (glCoreVertex2f x y tx ty)))
+        (let* ((cx (flo _cx)) (cy (flo _cy))
```
Comment (mgorges) [inline modules/ln_glcore/glcore.scm:370]: Note: Again changes that solely change indentation

File: modules/ln_glcore/glcore.scm [hunk 8]
```diff
@@ -338,24 +367,26 @@ SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
...
+               (r (flo _r)))
+          (glPushMatrix)
+          (glTranslatef cx cy 0.)
+          (glRotatef r 0. 0. 1.)
+          (_glCoreTextureBind t)
+          (glCoreBegin GL_TRIANGLE_STRIP)
+          (for-each
+           (lambda (p)
+             ;; TBD: should accept vectoralikes as point
+             (let* ((p (list->vector p))
```
Comment (mgorges) [inline modules/ln_glcore/glcore.scm:380]: Here is the actual change

**Comment by 0-8-15** [comment]
Am Wed, 04 Nov 2020 01:22:05 -0800
schrieb Matthias Görges <notifications@github.com>:

> @mgorges commented on this pull request.
> 
> 
> 
> > @@ -338,24 +367,26 @@ SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY
> > OF SUCH DAMAGE.  
>    (glcore:log 5 "glCoreTexturePolygonDraw")
>    (let ((entry (table-ref glCore:textures t #f)))
>      (if entry
> -      (let* ((cx (flo _cx)) (cy (flo _cy))
> -             (r (flo _r)))     
> -        (glPushMatrix)
> -        (glTranslatef cx cy 0.)
> -        (glRotatef r 0. 0. 1.)
> -        (_glCoreTextureBind t)
> -        (glCoreBegin GL_TRIANGLE_STRIP)
> -        (for-each (lambda (p) 
> -            (let* ((x (fl- (car p) cx))
> -                   (y (fl- (cadr p) cy))
> -                   (tx (caddr p))
> -                   (ty (cadddr p)))
> -              (glCoreVertex2f x y tx ty)))
> +        (let* ((cx (flo _cx)) (cy (flo _cy))
> 
> Note: Again changes that solely change indentation

I don't think so: This is one of the actual changes: replace repeated
traversal of the list `p` (which is O(n) in the length of the number
of `d`s in the `c`*X*/`r`) with vector-ref, which is constant time.

**Comment by 0-8-15** [comment]
> Note: Again changes that solely change indentation

Yes, sorry.

LISP is like Python:  You don't mess with the whitespace beyond the
traditional coding rules.

Otherwise... there is a difference: With Python you code is no longer
correct, with LISP it's still valid but automatic pretty printing will
kill your edits.

**Comment by mgorges** [comment]
> Note: Again changes that solely change indentation
> Yes, sorry. LISP is like Python: You don't mess with the whitespace beyond the traditional coding rules. Otherwise... there is a difference: With Python you code is no longer correct, with LISP it's still valid but automatic pretty printing will kill your edits.

The problem is with Git and the way it identifies changes as a workflow issue, not a code issues. - It is difficult to trace back why we changed what when when in the end the last change is a space addition one. Thus I might do some changes to the indentation just to prevent that before I merge it. Don't think there is a good automated solution for this though.

**Pull request merged**